### PR TITLE
Generate params for ForceTorqueSensorBroadcaster

### PR DESCRIPTION
**Pull request by tylerjw**
Signed-off-by: Tyler Weaver <tyler@picknik.ai>

Another PR to adapt a controller to use GenerateParameterLibrary

### CODE REVIEW

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp [hunk 1]
```diff
@@ -53,18 +47,19 @@ controller_interface::CallbackReturn ForceTorqueSensorBroadcaster::on_init()
...
-  interface_names_[1] = get_node()->get_parameter("interface_names.force.y").as_string();
-  interface_names_[2] = get_node()->get_parameter("interface_names.force.z").as_string();
-  interface_names_[3] = get_node()->get_parameter("interface_names.torque.x").as_string();
-  interface_names_[4] = get_node()->get_parameter("interface_names.torque.y").as_string();
-  interface_names_[5] = get_node()->get_parameter("interface_names.torque.z").as_string();
+  if (!param_listener_)
+  {
+    RCLCPP_ERROR(get_node()->get_logger(), "Parameters listener failed to construct in on_init");
+    return controller_interface::CallbackReturn::ERROR;
+  }
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:54]: Is this really needed here? Shouldn't line 36 already fail?

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp [hunk 2]
```diff
@@ -82,24 +77,24 @@ controller_interface::CallbackReturn ForceTorqueSensorBroadcaster::on_configure(
...
+  if (!params_.sensor_name.empty())
   {
     force_torque_sensor_ = std::make_unique<semantic_components::ForceTorqueSensor>(
-      semantic_components::ForceTorqueSensor(sensor_name_));
+      semantic_components::ForceTorqueSensor(params_.sensor_name));
   }
   else
   {
+    auto const & force = params_.interface_names.force;
+    auto const & torque = params_.interface_names.torque;
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:94]:
```suggestion
    auto const & force_names = params_.interface_names.force;
    auto const & torque_names = params_.interface_names.torque;
```

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp [hunk 3]
```diff
@@ -82,24 +77,24 @@ controller_interface::CallbackReturn ForceTorqueSensorBroadcaster::on_configure(
...
     force_torque_sensor_ = std::make_unique<semantic_components::ForceTorqueSensor>(
       semantic_components::ForceTorqueSensor(
-        interface_names_[0], interface_names_[1], interface_names_[2], interface_names_[3],
-        interface_names_[4], interface_names_[5]));
+        force.x, force.y, force.z, torque.x, torque.y, torque.z));
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:97]:
```suggestion
        force_names.x, force_names.y, force_names.z, torque_names.x, torque_names.y, torque_names.z));
```

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml [hunk 4]
```diff
@@ -0,0 +1,36 @@
+force_torque_sensor_broadcaster:
+  frame_id: {
+    type: string,
+    default_value: "",
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml:4]:
```suggestion
    default_value: "Sensor's frame_id in which values are published.",
```

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  }
+  sensor_name: {
+    type: string,
+    default_value: "",
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml:8]:
```suggestion
    default_value: "Name of the sensor used as prefix for interfaces if there are no individual interface names defined.",
```

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  }
+  interface_names:
+    force:
+      x: {
+        type: string,
+        default_value: "",
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml:14]:
```suggestion
        default_value: "Name of the state interface with force values on 'x' axis.",
```

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+      }
+      y: {
+        type: string,
+        default_value: "",
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml:18]:
```suggestion
        default_value: "Name of the state interface with force values on 'y' axis.",
```

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      }
+      z: {
+        type: string,
+        default_value: "",
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml:22]:
```suggestion
        default_value: "Name of the state interface with force values on 'z' axis.",
```

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+      }
+    torque:
+      x: {
+        type: string,
+        default_value: "",
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml:27]:
```suggestion
        default_value: "Name of the state interface with torque values around 'x' axis.",
```

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+      }
+      y: {
+        type: string,
+        default_value: "",
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml:31]:
```suggestion
        default_value: "Name of the state interface with torque values around 'y' axis.",
```

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+      }
+      z: {
+        type: string,
+        default_value: "",
```
Comment (destogl) [inline force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster_parameters.yaml:35]:
```suggestion
        default_value: "Name of the state interface with torque values around 'z' axis.",
```

**Review by destogl — CHANGES_REQUESTED** [review]
Few proposals to make this even more better! Thanks for all the work!

File: force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp (hunk 1 above)
Comment (tylerjw) [reply force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:54]: It guards against on_configure being called and resulting in a segfault in the case where on_init failed and param_listener_ is not yet valid.

Comment (destogl) [reply force_torque_sensor_broadcaster/src/force_torque_sensor_broadcaster.cpp:54]:
```suggestion
```

**Review by destogl — APPROVED** [review]

File: diff_drive_controller/src/diff_drive_controller.cpp [hunk 12]
```diff
@@ -423,7 +423,8 @@ controller_interface::CallbackReturn DiffDriveController::on_configure(
   {
     velocity_command_subscriber_ = get_node()->create_subscription<Twist>(
       DEFAULT_COMMAND_TOPIC, rclcpp::SystemDefaultsQoS(),
-      [this](const std::shared_ptr<Twist> msg) -> void {
+      [this](const std::shared_ptr<Twist> msg) -> void
+      {
```
Comment (erickisos) [inline diff_drive_controller/src/diff_drive_controller.cpp:427]:
```suggestion
      [this](const std::shared_ptr<Twist> msg) -> void {
```

File: diff_drive_controller/src/diff_drive_controller.cpp [hunk 13]
```diff
@@ -446,7 +447,8 @@ controller_interface::CallbackReturn DiffDriveController::on_configure(
     velocity_command_unstamped_subscriber_ =
       get_node()->create_subscription<geometry_msgs::msg::Twist>(
         DEFAULT_COMMAND_UNSTAMPED_TOPIC, rclcpp::SystemDefaultsQoS(),
-        [this](const std::shared_ptr<geometry_msgs::msg::Twist> msg) -> void {
+        [this](const std::shared_ptr<geometry_msgs::msg::Twist> msg) -> void
+        {
```
Comment (erickisos) [inline diff_drive_controller/src/diff_drive_controller.cpp:451]:
```suggestion
        [this](const std::shared_ptr<geometry_msgs::msg::Twist> msg) -> void {
```

File: diff_drive_controller/src/diff_drive_controller.cpp [hunk 14]
```diff
@@ -595,7 +597,8 @@ controller_interface::CallbackReturn DiffDriveController::on_shutdown(
 
 void DiffDriveController::halt()
 {
-  const auto halt_wheels = [](auto & wheel_handles) {
+  const auto halt_wheels = [](auto & wheel_handles)
+  {
```
Comment (erickisos) [inline diff_drive_controller/src/diff_drive_controller.cpp:601]:
```suggestion
  const auto halt_wheels = [](auto & wheel_handles) {
```

File: diff_drive_controller/src/diff_drive_controller.cpp [hunk 15]
```diff
@@ -625,7 +628,8 @@ controller_interface::CallbackReturn DiffDriveController::configure_side(
     const auto interface_name = feedback_type();
     const auto state_handle = std::find_if(
       state_interfaces_.cbegin(), state_interfaces_.cend(),
-      [&wheel_name, &interface_name](const auto & interface) {
+      [&wheel_name, &interface_name](const auto & interface)
+      {
```
Comment (erickisos) [inline diff_drive_controller/src/diff_drive_controller.cpp:632]:
```suggestion
      [&wheel_name, &interface_name](const auto & interface) {
```

File: diff_drive_controller/src/diff_drive_controller.cpp [hunk 16]
```diff
@@ -638,7 +642,8 @@ controller_interface::CallbackReturn DiffDriveController::configure_side(
 
     const auto command_handle = std::find_if(
       command_interfaces_.begin(), command_interfaces_.end(),
-      [&wheel_name](const auto & interface) {
+      [&wheel_name](const auto & interface)
+      {
```
Comment (erickisos) [inline diff_drive_controller/src/diff_drive_controller.cpp:646]:
```suggestion
      [&wheel_name](const auto & interface) {
```

File: gripper_controllers/include/gripper_controllers/gripper_action_controller_impl.hpp [hunk 17]
```diff
@@ -238,9 +238,8 @@ controller_interface::CallbackReturn GripperActionController<HardwareInterface>:
 {
   auto position_command_interface_it = std::find_if(
     command_interfaces_.begin(), command_interfaces_.end(),
-    [](const hardware_interface::LoanedCommandInterface & command_interface) {
-      return command_interface.get_interface_name() == hardware_interface::HW_IF_POSITION;
-    });
+    [](const hardware_interface::LoanedCommandInterface & command_interface)
+    { return command_interface.get_interface_name() == hardware_interface::HW_IF_POSITION; });
```
Comment (erickisos) [inline gripper_controllers/include/gripper_controllers/gripper_action_controller_impl.hpp:242]:
```suggestion
    [](const hardware_interface::LoanedCommandInterface & command_interface) {
      return command_interface.get_interface_name() == hardware_interface::HW_IF_POSITION; 
    });
```

File: gripper_controllers/include/gripper_controllers/gripper_action_controller_impl.hpp [hunk 18]
```diff
@@ -256,9 +255,8 @@ controller_interface::CallbackReturn GripperActionController<HardwareInterface>:
   }
   const auto position_state_interface_it = std::find_if(
     state_interfaces_.begin(), state_interfaces_.end(),
-    [](const hardware_interface::LoanedStateInterface & state_interface) {
-      return state_interface.get_interface_name() == hardware_interface::HW_IF_POSITION;
-    });
+    [](const hardware_interface::LoanedStateInterface & state_interface)
+    { return state_interface.get_interface_name() == hardware_interface::HW_IF_POSITION; });
```
Comment (erickisos) [inline gripper_controllers/include/gripper_controllers/gripper_action_controller_impl.hpp:259]:
```suggestion
    [](const hardware_interface::LoanedStateInterface & state_interface) {
      return state_interface.get_interface_name() == hardware_interface::HW_IF_POSITION;
    });
```

File: gripper_controllers/include/gripper_controllers/gripper_action_controller_impl.hpp [hunk 19]
```diff
@@ -274,9 +272,8 @@ controller_interface::CallbackReturn GripperActionController<HardwareInterface>:
   }
   const auto velocity_state_interface_it = std::find_if(
     state_interfaces_.begin(), state_interfaces_.end(),
-    [](const hardware_interface::LoanedStateInterface & state_interface) {
-      return state_interface.get_interface_name() == hardware_interface::HW_IF_VELOCITY;
-    });
+    [](const hardware_interface::LoanedStateInterface & state_interface)
+    { return state_interface.get_interface_name() == hardware_interface::HW_IF_VELOCITY; });
```
Comment (erickisos) [inline gripper_controllers/include/gripper_controllers/gripper_action_controller_impl.hpp:276]:
```suggestion
    [](const hardware_interface::LoanedStateInterface & state_interface) {
      return state_interface.get_interface_name() == hardware_interface::HW_IF_VELOCITY;
    });
```

File: joint_state_broadcaster/src/joint_state_broadcaster.cpp [hunk 20]
```diff
@@ -123,7 +123,8 @@ controller_interface::CallbackReturn JointStateBroadcaster::on_configure(
       "Publishing state interfaces defined in 'joints' and 'interfaces' parameters.");
   }
 
-  auto get_map_interface_parameter = [&](const std::string & interface) {
+  auto get_map_interface_parameter = [&](const std::string & interface)
+  {
```
Comment (erickisos) [inline joint_state_broadcaster/src/joint_state_broadcaster.cpp:127]:
```suggestion
  auto get_map_interface_parameter = [&](const std::string & interface) {
```

File: tricycle_controller/src/tricycle_controller.cpp [hunk 21]
```diff
@@ -370,7 +370,8 @@ CallbackReturn TricycleController::on_configure(const rclcpp_lifecycle::State &
   {
     velocity_command_subscriber_ = get_node()->create_subscription<TwistStamped>(
       DEFAULT_COMMAND_TOPIC, rclcpp::SystemDefaultsQoS(),
-      [this](const std::shared_ptr<TwistStamped> msg) -> void {
+      [this](const std::shared_ptr<TwistStamped> msg) -> void
+      {
```
Comment (erickisos) [inline tricycle_controller/src/tricycle_controller.cpp:374]:
```suggestion
      [this](const std::shared_ptr<TwistStamped> msg) -> void {
```

File: tricycle_controller/src/tricycle_controller.cpp [hunk 22]
```diff
@@ -392,7 +393,8 @@ CallbackReturn TricycleController::on_configure(const rclcpp_lifecycle::State &
   {
     velocity_command_unstamped_subscriber_ = get_node()->create_subscription<Twist>(
       DEFAULT_COMMAND_TOPIC, rclcpp::SystemDefaultsQoS(),
-      [this](const std::shared_ptr<Twist> msg) -> void {
+      [this](const std::shared_ptr<Twist> msg) -> void
+      {
```
Comment (erickisos) [inline tricycle_controller/src/tricycle_controller.cpp:397]:
```suggestion
      [this](const std::shared_ptr<Twist> msg) -> void {
```

File: tricycle_controller/src/tricycle_controller.cpp [hunk 23]
```diff
@@ -559,7 +561,8 @@ CallbackReturn TricycleController::get_traction(
   // Lookup the velocity state interface
   const auto state_handle = std::find_if(
     state_interfaces_.cbegin(), state_interfaces_.cend(),
-    [&traction_joint_name](const auto & interface) {
+    [&traction_joint_name](const auto & interface)
+    {
```
Comment (erickisos) [inline tricycle_controller/src/tricycle_controller.cpp:565]:
```suggestion
    [&traction_joint_name](const auto & interface) {
```

File: tricycle_controller/src/tricycle_controller.cpp [hunk 24]
```diff
@@ -574,7 +577,8 @@ CallbackReturn TricycleController::get_traction(
   // Lookup the velocity command interface
   const auto command_handle = std::find_if(
     command_interfaces_.begin(), command_interfaces_.end(),
-    [&traction_joint_name](const auto & interface) {
+    [&traction_joint_name](const auto & interface)
+    {
```
Comment (erickisos) [inline tricycle_controller/src/tricycle_controller.cpp:581]:
```suggestion
    [&traction_joint_name](const auto & interface) {
```

File: tricycle_controller/src/tricycle_controller.cpp [hunk 25]
```diff
@@ -599,7 +603,8 @@ CallbackReturn TricycleController::get_steering(
   // Lookup the velocity state interface
   const auto state_handle = std::find_if(
     state_interfaces_.cbegin(), state_interfaces_.cend(),
-    [&steering_joint_name](const auto & interface) {
+    [&steering_joint_name](const auto & interface)
+    {
```
Comment (erickisos) [inline tricycle_controller/src/tricycle_controller.cpp:607]:
```suggestion
    [&steering_joint_name](const auto & interface) {
```

File: tricycle_controller/src/tricycle_controller.cpp [hunk 26]
```diff
@@ -614,7 +619,8 @@ CallbackReturn TricycleController::get_steering(
   // Lookup the velocity command interface
   const auto command_handle = std::find_if(
     command_interfaces_.begin(), command_interfaces_.end(),
-    [&steering_joint_name](const auto & interface) {
+    [&steering_joint_name](const auto & interface)
+    {
```
Comment (erickisos) [inline tricycle_controller/src/tricycle_controller.cpp:623]:
```suggestion
    [&steering_joint_name](const auto & interface) {
```

**Review by erickisos — APPROVED** [review]
I left some comments to help you fix some of the linter issues in case your local configuration doesn't match the expected format.

👏

**Comment by destogl** [comment]
> I left some comments to help you fix some of the linter issues in case your local configuration doesn't match the expected format.

@erickisos thanks! But this was unfortunately not the issue with formatting. Now we should have everything correct.

**Comment by tylerjw** [comment]
I am developing on ubuntu 22.04 with Humble and/or Rolling and I believe I have all the required linters correctly installed. Is there a guide somewhere?

It seems that CI is not catching changes to formatting and I have submitted several small PRs to fix formatting since I started working on this change.

**Review by bmagyar — APPROVED** [review]

**Pull request merged**